collector/physical_disk.go

```diff
@@ -0,0 +1,269 @@
+	ReadBytesTotal   *prometheus.Desc
+	ReadsTotal       *prometheus.Desc
+	WriteBytesTotal  *prometheus.Desc
+	WritesTotal      *prometheus.Desc
+	ReadTime         *prometheus.Desc
+	WriteTime        *prometheus.Desc
+	IdleTime         *prometheus.Desc
+	SplitIOs         *prometheus.Desc
+	ReadLatency      *prometheus.Desc
+	WriteLatency     *prometheus.Desc
+	ReadWriteLatency *prometheus.Desc
+
+	diskWhitelistPattern *regexp.Regexp
+	diskBlacklistPattern *regexp.Regexp
+}
+
+// NewPhysicalDiskCollector ...
+func NewPhysicalDiskCollector() (Collector, error) {
+	const subsystem = "physical_disk"
+
+	return &PhysicalDiskCollector{
+		RequestsQueued: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "requests_queued"),
+			"The number of requests queued to the disk (PhysicalDisk.CurrentDiskQueueLength)",
+			[]string{"disk"},
+			nil,
+		),
+
+		ReadBytesTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "read_bytes_total"),
+			"The number of bytes transferred from the disk during read operations (PhysicalDisk.DiskReadBytesPerSec)",
+			[]string{"disk"},
+			nil,
+		),
+
+		ReadsTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "reads_total"),
+			"The number of read operations on the disk (PhysicalDisk.DiskReadsPerSec)",
+			[]string{"disk"},
+			nil,
+		),
+
+		WriteBytesTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "write_bytes_total"),
+			"The number of bytes transferred to the disk during write operations (PhysicalDisk.DiskWriteBytesPerSec)",
+			[]string{"disk"},
+			nil,
+		),
+
+		WritesTotal: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "writes_total"),
+			"The number of write operations on the disk (PhysicalDisk.DiskWritesPerSec)",
+			[]string{"disk"},
+			nil,
+		),
+
+		ReadTime: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "read_seconds_total"),
+			"Seconds that the disk was busy servicing read requests (PhysicalDisk.PercentDiskReadTime)",
+			[]string{"disk"},
+			nil,
+		),
+
+		WriteTime: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "write_seconds_total"),
+			"Seconds that the disk was busy servicing write requests (PhysicalDisk.PercentDiskWriteTime)",
+			[]string{"disk"},
+			nil,
+		),
+
+		IdleTime: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "idle_seconds_total"),
+			"Seconds that the disk was idle (PhysicalDisk.PercentIdleTime)",
+			[]string{"disk"},
+			nil,
+		),
+
+		SplitIOs: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "split_ios_total"),
+			"The number of I/Os to the disk were split into multiple I/Os (PhysicalDisk.SplitIOPerSec)",
+			[]string{"disk"},
+			nil,
+		),
+
+		ReadLatency: prometheus.NewDesc(
+			prometheus.BuildFQName(Namespace, subsystem, "read_latency_seconds_total"),
+			"Shows the average time, in seconds, of a read operation from the disk (PhysicalDisk.AvgDiskSecPerRead)",
```

Comment:
This (and for `WriteLatency` as well) is actually total latency for all reads, correct?

Reply:
Not sure, the MSFT description says "Average time, in seconds, of a read operation of data from the disk.". I think this is averaged against the previous sample, not in total. But again, I'm not sure.